lib/kb_Msuite/kb_MsuiteImpl.py

```diff
@@ -128,6 +129,49 @@ def run_checkM_lineage_wf(self, ctx, params):
                              'result is not type dict as required.')
         # return the results
         return [result]
+
+    def lineage_wf(self, ctx, params):
+        """
+        A "local method" for calling lineage_wf directly.
+        :param params: instance of type "LineageWfParams" (* * Parameters for
+           lineage_wf, which runs as a "local method". * * Required
+           arguments: *   bin_dir - required - Path to the directory where
+           your bins are located *   out_dir - required - Path to a directory
+           where we will write output files *   options - optional - A
+           mapping of options to pass to lineage_wf. See the README.md *    
+           in the kb_Msuite repo for a list of all of these. For options that
+           have no value, simply *     pass an empty string.) -> structure:
+           parameter "bin_dir" of String, parameter "out_dir" of String,
+           parameter "options" of mapping from String to String
+        :returns: instance of type "LineageWfResult" (* * Output results of
+           running the lineage_wf local method. * We simply give the raw
+           standard out from checkm, which may be an error message (checkm
+           does * not use stderr) * * Fields: *   stdout - The standard out
+           given by checkm after running.) -> structure: parameter "stdout"
+           of String
+        """
+        # ctx is the context object
+        # return variables are: result
+        #BEGIN lineage_wf
+        assert params['log_path'], 'You must provide the path for a log file ("log_path")'
+        assert params['input_dir'], 'You must provide an input directory ("input_dir")'
+        assert params['output_dir'], 'You must provide an output directory ("output_dir")'
+        assert os.path.isdir(params['input_dir']), 'The input directory does not exist'
+        assert os.path.isdir(params['output_dir']), 'The output directory does not exist'
+        assert os.listdir(params['output_dir']) == [], 'The output directory must be empty'
+        in_dir = params['input_dir']
+        out_dir = params['output_dir']
+        log_path = params['log_path']
+        run_checkm(in_dir, out_dir, log_path, params.get('options'))
+        result = {}
```

Comment:
Should run_checkm return a stdout string? Same string should be returned in lineage_wf result.

Reply:
It writes all stdout and stderr into a file in log_path. In several cases I found that stdout is going to be extremely long, thousands of lines, so I think people might not want to have it in a string / in memory by default.

The downside is it's not user friendly to have the user open up a log file to see any errors. But also it's not easy for us to automatically know if there was an error -- checkm uses stdout for many errors and does not use error codes.

Reply:
👍  Could you change spec file to reflect the method is returning an empty dict?